File: DataStructureImplementationPart1/PriorityQueue.java
package DataStructureImplementationPart1;
import java.util.*;
import java.util.ArrayList;

public class PriorityQueue {
    static class Pair{
        String value;
        int priority;
        Pair(String value,int priority){
            this.value = value;
            this.priority = priority; 
        }
    }
    private ArrayList<Pair> arr;

    public PriorityQueue(){
        arr = new ArrayList<>();
    }


    public String top(){
        if(arr.size() ==0){
           throw new IllegalStateException("heap empty");
        }
         return arr.get(0).value;
        
    }

    public void insert(String x, int weight){
        Pair tempPair = new Pair(x,weight);
        arr.add(tempPair);
        heapifyUp(arr.size()-1);
    }

    public void remove(){
        if (arr.size() == 0) {
            throw new IllegalStateException("heap empty");
        }
        arr.set(0,arr.get(arr.size()-1));
        arr.remove(arr.size()-1);
        if(arr.size()>0){
              heapifyDown(0);
        }
      
    }
    private void heapifyDown(int index){
        int left = 2 * index + 1;
        int right = 2 * index + 2;
        int largest = index;

    if (left < arr.size() && arr.get(left).priority > arr.get(largest).priority) {
        largest = left;
    }

    if (right < arr.size() && arr.get(right).priority > arr.get(largest).priority) {
        largest = right;
    }

    if (largest != index) {
        swap(index, largest);
        heapifyDown(largest);
    }
        
    }
    private void heapifyUp(int index){
        if(index==0){
            return;
        }
        int parent  = (index-1)/2;
        if(arr.get(parent).priority < arr.get(index).priority){
            swap(parent,index);
            heapifyUp(parent);
        }
    }
    
    private void swap(int index1, int index2){
        Pair temp = arr.get(index1);
        arr.set(index1,arr.get(index2));
        arr.set(index2,temp);
    }
}

File: homework3/priorityqueue.py
# Technique: Array-based Max Heap storing 
# Time Complexity: insert O(log n), remove O(log n), top O(1)
# Space Complexity: O(n)


class PriorityQueue:
    def __init__(self):
        self.arr = []

    def top(self):
        if not self.arr:
            return None
        return self.arr[0][1]  

    def insert(self, x, weight):
        self.arr.append((-weight, x))
        self._bubbleUp(len(self.arr) - 1)

    def remove(self):
        # removes highest priority (root) element
        if not self.arr:
            return
        self.arr[0] = self.arr[-1]
        self.arr.pop()
        if self.arr:
            self._bubbleDown(0)

    def _bubbleUp(self, idx):
        parent = (idx - 1) // 2
        while idx > 0 and self.arr[idx][0] < self.arr[parent][0]:
            self.arr[idx], self.arr[parent] = self.arr[parent], self.arr[idx]
            idx = parent
            parent = (idx - 1) // 2

    def _bubbleDown(self, idx):
        n = len(self.arr)
        while True:
            smallest = idx
            left  = 2 * idx + 1
            right = 2 * idx + 2

            if left < n and self.arr[left][0] < self.arr[smallest][0]:
                smallest = left
            if right < n and self.arr[right][0] < self.arr[smallest][0]:
                smallest = right

            if smallest == idx:
                break

            self.arr[idx], self.arr[smallest] = self.arr[smallest], self.arr[idx]
            idx = smallest


# Test 1: basic inserts, top returns highest priority
pq = PriorityQueue()
pq.insert("Cricket", 10)
pq.insert("Tabletennis", 9)
pq.insert("Football", 7)
pq.insert("Movies", 5)
print(pq.top())    

# Test 2: remove in priority order
pq.remove()
print(pq.top())    
pq.remove()
print(pq.top())    
pq.remove()
print(pq.top())    
pq.remove()
print(pq.top())    

# Test 3: single element
pq2 = PriorityQueue()
pq2.insert("only task", 1)
print(pq2.top())   
pq2.remove()
print(pq2.top())   

# Test 4: same priority
pq3 = PriorityQueue()
pq3.insert("task a", 5)
pq3.insert("task b", 5)
print(pq3.top())   # task a or task b (either valid)

# Time spent: ~20 minutes